libretro jni builid failed

### Platform

libretro / Retroarch

### Compiler and build tool versions

ndk 21e

### Operating system version

ubuntu 20.04

### Build commands used

$NDK/ndk-build

### What happens

[armeabi-v7a] Compile++ thumb: retro <= ControlMapper.cpp
[armeabi-v7a] Compile++ thumb: retro <= TextureReplacer.cpp
/home/xielei/ppsspp/libretro/jni/../../Core/TextureReplacer.cpp:645:19: error: no matching function for call to 'zip_fread'
                        pngdata.resize(zip_fread(zf, pngdata.data(), pngdata.size()));
                                       ^~~~~~~~~
../../ext/libzip/zip.h:409:24: note: candidate function not viable: no known conversion from 'const std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >::value_type *' (aka 'const char *') to 'void *' for 2nd argument
ZIP_EXTERN zip_int64_t zip_fread(zip_file_t *_Nonnull, void *_Nonnull, zip_uint64_t);
                       ^
/home/xielei/ppsspp/libretro/jni/../../Core/TextureReplacer.cpp:1171:20: error: no matching function for call to 'zip_fread'
                                pngdata.resize(zip_fread(zf, pngdata.data(), pngdata.size()));
                                               ^~~~~~~~~
../../ext/libzip/zip.h:409:24: note: candidate function not viable: no known conversion from 'const std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >::value_type *' (aka 'const char *') to 'void *' for 2nd argument
ZIP_EXTERN zip_int64_t zip_fread(zip_file_t *_Nonnull, void *_Nonnull, zip_uint64_t);
                       ^
2 errors generated.
make: *** [/home/xielei/ndk/android-ndk-r21e/build/core/build-binary.mk:478: /home/xielei/ppsspp/libretro/obj/local/armeabi-v7a/objs/retro/__/__/Core/TextureReplacer.o] Error 1

### PPSSPP version affected

commit e97d5498c639919c4c533bb55a8a59fb50a87373

### Last working version

_No response_

### Checklist

- [X] Test in the [latest git build](https://buildbot.orphis.net/ppsspp/) in case it's already fixed.
- [X] Make sure to run `git submodule update --init --recursive` before building.
- [X] [Search for other reports](https://github.com/hrydgard/ppsspp/search?q=my+issue&type=issues) of the same issue.
- [X] Try deleting the build directory and running the build again.
- [X] Check GitHub Actions, which builds every merge and PR.
- [X] Include logs and help us reproduce.

--- a/Core/TextureReplacer.cpp
+++ b/Core/TextureReplacer.cpp
@@ -642,10 +642,10 @@ bool TextureReplacer::PopulateLevelFromZip(ReplacedTextureLevel &level, bool ign
 		if (zsize != INVALID_ZIP_SIZE)
 			pngdata.resize(zsize);
 		if (!pngdata.empty()) {
-			pngdata.resize(zip_fread(zf, pngdata.data(), pngdata.size()));
+			pngdata.resize(zip_fread(zf, &pngdata[0], pngdata.size()));
 		}
 
-		if (png_image_begin_read_from_memory(&png, pngdata.data(), pngdata.size())) {
+		if (png_image_begin_read_from_memory(&png, &pngdata[0], pngdata.size())) {
 			// We pad files that have been hashrange'd so they are the same texture size.
 			level.w = png.width;
 			level.h = png.height;
@@ -1168,10 +1168,10 @@ void ReplacedTexture::PrepareData(int level) {
 			if (zsize != INVALID_ZIP_SIZE)
 				pngdata.resize(zsize);
 			if (!pngdata.empty()) {
-				pngdata.resize(zip_fread(zf, pngdata.data(), pngdata.size()));
+				pngdata.resize(zip_fread(zf, &pngdata[0], pngdata.size()));
 			}
 
-			if (!png_image_begin_read_from_memory(&png, pngdata.data(), pngdata.size())) {
+			if (!png_image_begin_read_from_memory(&png, &pngdata[0], pngdata.size())) {
 				ERROR_LOG(G3D, "Could not load texture replacement info: %s - %s (zip)", info.file.c_str(), png.message);
 				cleanup();
 				return;

--- a/Core/Util/GameManager.cpp
+++ b/Core/Util/GameManager.cpp
@@ -325,6 +325,9 @@ bool GameManager::InstallGame(Path url, Path fileName, bool deleteAfter) {
 		if (DetectTexturePackDest(z, info.textureIniIndex, dest)) {
 			INFO_LOG(HLE, "Installing '%s' into '%s'", fileName.c_str(), dest.c_str());
 			File::CreateFullPath(dest);
+			// Install as a zip file if textures.ini is in the root.  Performs better on Android.
+			if (info.stripChars == 0)
+				return InstallMemstickZip(z, fileName, dest / "textures.zip", info, deleteAfter);
 			File::CreateEmptyFile(dest / ".nomedia");
 			return InstallMemstickGame(z, fileName, dest, info, true, deleteAfter);
 		} else {
@@ -615,6 +618,60 @@ bail:
 	}
 	SetInstallError(sy->T("Storage full"));
 	return false;
+}
+
+bool GameManager::InstallMemstickZip(struct zip *z, const Path &zipfile, const Path &dest, const ZipFileInfo &info, bool deleteAfter) {
+	size_t allBytes = 0;
+	size_t bytesCopied = 0;
+
+	auto sy = GetI18NCategory("System");
+
+	// We don't need the zip anymore, as we're going to copy it as-is.
+	zip_close(z);
+	z = nullptr;
+
+	// Not using File::Copy() so we can report progress.
+	FILE *inf = File::OpenCFile(zipfile, "rb");
+	if (!inf)
+		return false;
+
+	allBytes = (size_t)File::GetFileSize(inf);
+	FILE *outf = File::OpenCFile(dest, "wb");
+	if (!outf) {
+		SetInstallError(sy->T("Storage full"));
+		fclose(inf);
+		return false;
+	}
+
+	const size_t blockSize = 1024 * 128;
+	u8 *buffer = new u8[blockSize];
+	while (bytesCopied < allBytes) {
+		size_t readSize = std::min(blockSize, allBytes - bytesCopied);
+		if (fread(buffer, readSize, 1, inf) != 1)
+			break;
+		if (fwrite(buffer, readSize, 1, outf) != 1)
+			break;
+		bytesCopied += readSize;
+		installProgress_ = (float)allBytes / (float)allBytes;
+	}
+
+	delete[] buffer;
+	fclose(inf);
+	fclose(outf);
+
+	if (bytesCopied < allBytes) {
+		File::Delete(dest);
+		SetInstallError(sy->T("Storage full"));
+		return false;
+	}
+
+	installProgress_ = 1.0f;
+	if (deleteAfter) {
+		File::Delete(zipfile);
+	}
+	InstallDone();
+	ResetInstallError();
+	return true;
 }
 
 bool GameManager::InstallZippedISO(struct zip *z, int isoFileIndex, const Path &zipfile, bool deleteAfter) {

--- a/Core/Util/GameManager.h
+++ b/Core/Util/GameManager.h
@@ -79,6 +79,7 @@ public:
 private:
 	bool InstallGame(Path url, Path tempFileName, bool deleteAfter);
 	bool InstallMemstickGame(struct zip *z, const Path &zipFile, const Path &dest, const ZipFileInfo &info, bool allowRoot, bool deleteAfter);
+	bool InstallMemstickZip(struct zip *z, const Path &zipFile, const Path &dest, const ZipFileInfo &info, bool deleteAfter);
 	bool InstallZippedISO(struct zip *z, int isoFileIndex, const Path &zipfile, bool deleteAfter);
 	bool InstallRawISO(const Path &zipFile, const std::string &originalName, bool deleteAfter);
 	void InstallDone();

